docs/development/building-good-containers.md

```diff
@@ -0,0 +1,119 @@
+
+There are also systems which support running containers as `rootless`.
+In order to ensure that your containers can run on these system you
+want to avoid running process as root as they may end up depending
+on capabilities that are not available when they are run `rootless`.
+
+When using docker/podman you can change the user with the `USER` command.
+Well designed base containers will have already done this, but you should
+always check what user will be used by default.
+
+```shell
+USER 1001
+RUN chown -R 1001:0 /some/directory
+```
+
+It is recommended that you build your containers as `non-root`.
+
+* Avoid using trusted ports
+
+Ports below 1024 are considered `trusted` and a process must have
+additional priviledges to be able to bind to them.
+
+If you respect the recommendation to build your containers as
+`non-root` then your process will not be able to bind to the
+trusted ports.
+
+Further, although there are ways to set capabilities on the
+container to allow non root users to bind to these ports, at
+the time of this writing doing so may cause other problems such
+as not being able use the `NODE_EXTRA_CA_CERTS` environment
+variable. You can read more about the issue
+[here](https://github.com/nodejs/node/pull/37727).
+
+Typical work arounds which include using firewall functionality to
+proxy ports most often don't work in containers as the firewall
+capability is often not included in container distributions.
+
+It is recommended that you build your containers so that internally
+only ports 1024 and above are used.
+
+## Use multi-stage builds
+
+While layering can reduce the size of the layer that contains
+your Node.js application, the total size of all of the layers
+for a container still ends up being a concern. This is
+particularly true in Kubernetes environments where Pods and
+Nodes may come/go frequently and the case were all layers
+need to be installed is more common than might be expected.
+
+Best practice is to use a two stage build where a larger
+build image is used to build an application and then the
+resulting artifacts are copied to a run image. The build
+image includes all of the tools needed to build the application
+(compilers etc.) while the run image only includes what's
+needed for the application to run. The size difference
+between the build and run images can often be significant.
+
+There are different ways to build containers but most
+will support using a build and run image. You read more
+about a few examples in these references:
+
+[multistage-build](https://docs.docker.com/develop/develop-images/multistage-build/)
+[Modern web applications on OpenShift: Part 2 -- Using chained builds](https://developers.redhat.com/blog/2018/10/23/modern-web-applications-on-openshift-part-2-using-chained-builds)
+
+It is recommended that you use multi-stage builds to minimize
+container size.
+
+
+## Add key tools for problem determination
+
+It can be difficult to add tooling/packages after the fact when
```

Comment:
Are we suggesting any tooling/packages in the framework? If so, we can link them here.

Reply:
I don't think we have a list right now. That would be a good addition but I think we can add in a follow on unless somebody has a good list handy.